### PR TITLE
feat(ckb): Handle UTXO bound multiply type assets

### PR DESCRIPTION
**Pull request by duanyytop**
## Main Changes

- Collect all RGB++ type assets as CKB TX inputs for RGB++ assets transfer and leap, and only transfer and leap target RGB++ assets to the receiver, and transfer the other RGB++ assets to the sender himself/herself
- When a UTXO is binding spore and xUDT assets, the error will be thrown for RGB++ assets transfer and leap
- Use the single function to check RGB++ live cells and throw errors when the live cells are invalid, and add test cases for the function to make sure the checker function works

## L1 transfer on Testnet

BTC TX:  
https://mempool.space/testnet/tx/78a52773c210c149384bfef18491328474e16b61bc5eb7d414a295fff502087d

CKB TX: 
https://pudge.explorer.nervos.org/transaction/0x475b845ed3e946f59119cc5958031eae75b6cd1614c67b2660d80c6d4b3ce439

## L1 -> L2 on Testnet

BTC TX: 
https://mempool.space/testnet/tx/10869e223d02a0d4d27be4ec4091d95f84b6d90c9425de98d6bee826cee1c907

CKB TX: 
https://pudge.explorer.nervos.org/transaction/0xa5322bd75196e45adabb2599f650856ee83a82c2045299bc79ab68ccc2992ea7

### CODE REVIEW

**Review by ahonn — APPROVED** [review]

**Review by ShookLyngs — APPROVED** [review]

File: packages/ckb/src/utils/spore.ts [hunk 1]
```diff
@@ -95,3 +97,22 @@ export const generateSporeTransferCoBuild = (
...
+  if (!sporeCells || sporeCells.length === 0) {
+    throw new NoRgbppLiveCellError('No spore rgbpp cells found with the spore rgbpp lock args');
+  }
+  if (sporeCells.length > 1) {
+    throw new RgbppUtxoBindMultiTypeAssetsError('The BTC UTXO must not be bound to multiple CKB cells');
+  }
+  const sporeCell = sporeCells[0];
+
+  if (!sporeCell.output.type) {
+    throw new NoRgbppLiveCellError('The cell with the rgbpp lock args has no spore asset');
```
Comment (Flouse) [inline packages/ckb/src/utils/spore.ts:112]:
```suggestion
    throw new RgbppSporeTypeMismatchError('The cell with the rgbpp lock args has no spore asset');
```

**Review by Flouse — CHANGES_REQUESTED** [review]

File: packages/ckb/src/utils/rgbpp.ts [hunk 2]
```diff
@@ -219,3 +226,28 @@ export const throwErrorWhenTxInputsExceeded = (inputLen: number) => {
...
+  const typeCells = cells.filter((cell) => !!cell.output.type);
+  if (typeCells.length === 0) {
+    throw new NoRgbppLiveCellError('No rgbpp cells found with the rgbpp lock args');
+  }
+  const isSporeExist = typeCells.some((cell) => isScriptPartialEqual(cell.output.type!, getSporeTypeScript(isMainnet)));
+  if (isSporeExist) {
+    throw new RgbppUtxoBindMultiTypeAssetsError(
+      'The BTC UTXO must not be bound to Spore and xUDT cells at the same time',
+    );
+  }
```
Comment (Flouse) [inline packages/ckb/src/utils/rgbpp.ts:248]: Some nonUDTType and nonClusterSpore may exist.
I think using `isTypeAssetSupported` to check all the cells is more appropriate.

Comment (duanyytop) [reply packages/ckb/src/utils/rgbpp.ts]: I used the function `isUDTTypeSupported` to filter non-xUDT assets for RGB++ xUDT transfer and leap in https://github.com/ckb-cell/rgbpp-sdk/pull/139/commits/b009bd7baa34199cc3115588d8d107b3035c61a3

File: packages/ckb/src/utils/rgbpp.ts [hunk 3]
```diff
@@ -219,3 +225,31 @@ export const throwErrorWhenTxInputsExceeded = (inputLen: number) => {
...
+  isMainnet: boolean,
+) => {
+  if (!cells || cells.length === 0) {
+    throw new NoRgbppLiveCellError('No rgbpp cells found with the rgbpp lock args');
+  }
+  const typeCells = cells.filter((cell) => !!cell.output.type);
+  if (typeCells.length === 0) {
+    throw new NoRgbppLiveCellError('No rgbpp cells found with the rgbpp lock args');
+  }
+  const isuDTTypeNotSupported = typeCells.some(
```
Comment (Flouse) [inline packages/ckb/src/utils/rgbpp.ts:242]: typo: isuDTTypeNotSupported -> isUDTTypeNotSupported

Comment (duanyytop) [reply packages/ckb/src/utils/rgbpp.ts]: Fixed https://github.com/ckb-cell/rgbpp-sdk/pull/139/commits/5fd332e21251e1119ab90dbc297f5a19e1a73265

**Review by Flouse — APPROVED** [review]

**Pull request merged**